test/api.js

```diff
@@ -336,4 +336,52 @@ describe('API', () => {
 				.catch(done);
 		});
 	});
+
+	describe('output depth limit', () => {
+		const baseDir = path.join(__dirname, 'deep-dependencies');
+		const entryFile = path.join(baseDir, 'a.js');
+
+		it('limits the output depth to 0', async () => {
+			const res = await madge(entryFile, {depth: 0});
+
+			res.obj().should.eql({
+				'a.js': []
+			});
+		});
+
+		it('limits the output depth to 1', async () => {
+			const res = await madge(entryFile, {depth: 1});
+
+			res.obj().should.eql({
+				'a.js': ['b.js', 'c.js'],
+				'b.js': [],
+				'c.js': ['b.js', 'c.js']
+			});
+		});
+
+		it('limits the output depth to 2', async () => {
+			const res = await madge(entryFile, {depth: 2});
+
+			res.obj().should.eql({
+				'a.js': ['b.js', 'c.js'],
+				'b.js': [],
+				'c.js': ['d.js', 'e.js'],
+				'd.js': [],
+				'e.js': ['b.js', 'c.js']
+			});
+		});
+
+		it('limits the output depth to 3', async () => {
+			const res = await madge(entryFile, {depth: 3});
+
+			res.obj().should.eql({
+				'a.js': ['b.js', 'c.js'],
+				'b.js': [],
+				'c.js': ['d.js', 'e.js'],
+				'd.js': [],
+				'e.js': ['f.js'],
+				'f.js': ['b.js', 'c.js']
+			});
+		});
```

Comment:
I apologize for my recent lack of activity. I'm currently sorting out a few details, as I've noticed that the implementation of `C, E, F, C` with `depth=3` in the test file seems to induce a self-dependency, which isn't accounted for in the final test case (the highlighted one). I need a bit more time to fully understand this, as I initially thought this PR primarily involved output formatting.

Could you @keijokapp please provide a specific use case or example that illustrates the necessity of this functionality? It would greatly help in evaluating its impact and relevance.

Thank you for your understanding and support.

Reply:
I might not fully understand the question. 

> which isn't accounted for in the final test case (the highlighted one)

The output of `depth=3` does show `c -> e -> f -> c` dependency cycle (?) So do tests with `depth=1` and `depth=2`. In the context of this feature, there isn't anything special about circular dependencies. They are handled just like any other "back-references" (ie references from deep dependencies to non-deep dependencies). If there are circular dependencies in the output, it means there were circular dependency in the original graph. These tests make sure that they are handled (short-circuited) correctly.

> please provide a specific use case or example that illustrates the necessity of this functionality

You mean "circular dependencies"? I don't think I'm the right person to explain the use cases for circular dependencies because I personally have a strong distaste for them. But they are nevertheless common and in rare cases not easily avoidable. This tool specifically handles them eg. by coloring them red on the dot graph. The feature doesn't have anything specifically to do with circular dependencies.

---

When it comes to "limiting the depth", there are two ways to go about it. One is to stop traversing at `depth` and ignore all references originating from that point onward. This is not usually what the user wants. Not showing those "back-references" gives a wrong impression about the dependency structure. For example, at `depth=1`, the user might see `b.js` and `c.js` as fully independent branches while, in fact, `b.js` is an indirect dependency of `c.js`.

Instead, as noted in the original submission, this feature traverses through the whole tree and does not hide those indirect relationships. It only removes the intermediate deep nodes.

> I initially thought this PR primarily involved output formatting.

It really is just about the output. It is to remove the "noise" while keeping the useful relationship information. It allows the user to focus around the close proximity of specific modules. It doesn't change anything else about the tool (eg. module resolution, parsing, traversing).

Reply:
Thank you for your detailed response, and I appreciate your patience as I navigate through this PR. After multiple reviews, I'm still grappling with some aspects of the implementation and would like to discuss these further.

```markdown
1. **File Structure and Depth Concerns**
   My apologies for any confusion caused by the `C-D-E-C` example I initially suggested—it may not have been the best illustration of the concept. Here’s a refresher on the file structure from our tests:

   ```text
   dependencies = {
       'a.js': ['b.js', 'c.js'],
       'b.js': [],
       'c.js': ['d.js', 'e.js'],
       'd.js': [],
       'e.js': ['f.js'],
       'f.js': ['g.js', 'c.js'],
       'g.js': ['b.js']
   }

   a.js
   ├─ b.js
   └─ c.js
      ├─ d.js
      └─ e.js
         └─ f.js
            ├─ g.js
            │  └─ b.js
            └─ c.js (circular dependency back to c.js)
   ```

   This tree structure helps illustrate potential dependency chains, such as `A-C-E-F` is like the `A => B => C => D` mentioned in the PR description. When discussing `depth=1`, I find that the outputs might not align with what one would expect, like `A-C-D` in the example in the description, which would be `A-E-F` in our test case accordingly. This which should help us better understand the effects of limiting depth. Could we consider revisiting this part to ensure it aligns more clearly with our objectives?


2. **Circular Dependencies**
   I generally strive to avoid circular dependencies, too. I found this repository trying to include checks for these circular dependencies in the CI/CD process. Kudos to the contributor for implementing such a valuable feature!

3. **Clarification on the 'depth' Functionality**
   Could you possibly provide more concrete examples or use cases for the `depth` functionality? While your explanation was insightful, additional examples would help solidify my understanding of its practical impact.

4. **Depth Traversal Options**
   Concerning your point on depth traversal:
   > When it comes to "limiting the depth", there are two ways to go about it. One is to stop traversing at `depth` and ignore all references originating from that point onward. This is not usually what the user wants.
   
   I slightly disagree here. In some scenarios, particularly during refactoring, it could be beneficial to understand the breadth of a module's direct influence, which would support limiting traversal as a feature.

5. **Confusion Over Terminology**
   These two terms in the description seem to mean the same thing, leading to some confusion. Could you clarify these definitions?
  - > the closest input module ("deep dependencies")
  - > non-deep modules (i.e. modules that are <=depth steps away from any input module)

6. **Implementation Details**
   I've noticed significant changes in the `convertTree(depTree, tree, pathCache)` method, particularly the removal of `pathCache`. This alteration could potentially lead to increase execution times significantly. I believe reintroducing `pathCache` might enhance performance. I'll need some more time to evaluate this change thoroughly.

Thank you once again for your support and understanding as we refine this feature.

Reply:
I added a few nodes for the tree so that we can be on the same page with "deep dependencies"

```text
dependencies = {
    'a.js': ['b.js', 'c.js'],
    'b.js': ['h.js'],
    'c.js': ['d.js', 'e.js'],
    'd.js': [],
    'e.js': ['f.js'],
    'f.js': ['g.js', 'c.js'],
    'g.js': ['b.js', 'h.js'],
    'h.js': ['i.js'],
    'i.js': ['j.js'],
    'j.js': ['k.js'],
    'k.js': ['l.js'],
    'l.js': []
}

a.js
├─ b.js
│  └─ h.js
│     └─ i.js
│        └─ j.js
│           └─ k.js
│              └─ l.js
└─ c.js
   ├─ d.js
   └─ e.js
      └─ f.js
         ├─ g.js
         │  ├─ b.js
         │  └─ h.js
         │     └─ i.js
         │        └─ j.js
         │           └─ k.js
         │              └─ l.js
         └─ c.js (circular dependency back to c.js)
```

For my understanding, with `depth=1`, we are excluding file `b`, `c`. And with `depth=2` we are also excluding `h`, `d` and `e` in addition. But here `h` would appear again in the chain `ACEFGH` which made the output unclear.
(This also shows that removing the `pathCache` might be a necessity to implement this feature, depending on what we are trying to achieve.)